Add support for RedHat 9

kibahop opened the pull request:
This adds so far minimally tested support for RedHat 9. The most important change in IPA version 4.10, that is atm in RedHat 9, is that mod_nss is replaced with with mod_ssl for the WebUI. Installer will also clash if UID_MAX and GID_MAX in login.defs overlap with idstart. There is a new parameters to handle that 'adjust_login_defs'. By default this is false.

kibahop committed:
Add support for RedHat 9

Signed-off-by: Kibahop <petri.lammi@puppeteers.net>

mattock commented:
The GitHub Action failures are false positives. I added some commits on top that fix a few genuine issues, but the rest is just GHA runtime environment issues. The first and possibly the last is this one:

* https://github.com/voxpupuli/gha-puppet/issues/23

mattock commented on manifests/config/webui.pp:
```diff
@@ -18,33 +18,37 @@
+    # IPA switched to mod_ssl as the crypto engine for Apache as of version 4.7.0
+    # see https://www.freeipa.org/page/Releases/4.7.0#Highlights_in_4.7.0
+    if versioncmp($facts['ipa_server_version'], '4.7.0') < 0 {
+      exec { 'semanage-port-http_port_t':
```
Does this whole block become unnecessary with FreeIPA > 4.7.0? Most of this does not look specific to a crypto library change.

mattock commented on manifests/init.pp:
```diff
@@ -240,7 +244,24 @@
+          ;
+        'adjust uid max':
+          line  => "UID_MAX\t11999",
+          match => '^UID_MAX\s*60000$',
```
Perhaps the match should be more flexible, e.g. **^UID_MAX\s*\d*$**. That way if the value is not exactly 60000 file_line would not fail or worse, add additional lines because no match is was found.

mattock replied:
Also, does this apply to both FreeIPA clients and servers?

mattock commented on manifests/init.pp:
```diff
@@ -240,7 +244,24 @@
+          ;
+        'adjust gid max':
+          line  => "GID_MAX\t11999",
+          match => '^GID_MAX\s*60000$',
```
Same comment here as above.

mattock commented on manifests/install/server.pp:
```diff
@@ -16,6 +16,32 @@
+  # Note: SUB_* only affect user/group mapping in containers, so not of
+  # concern here
+  if $easy_ipa::adjust_login_defs {
+    if $easy_ipa::idstart < Integer($facts['uid_max']) {
```
I would just return an Integer from Ruby. Easier, cleaner and more intuitive.

mattock commented on manifests/install/server.pp:
```diff
@@ -16,6 +16,32 @@
+    if $easy_ipa::idstart < Integer($facts['uid_max']) {
+      $uid_max_value = $easy_ipa::idstart -1
+    }
+    if $easy_ipa::idstart < Integer($facts['gid_max']) {
```
Same here.

mattock commented on lib/facter/gid_max.rb:
```diff
@@ -0,0 +1,8 @@
+Facter.add(:gid_max) do
+  setcode do
+    lines = File.readlines("/etc/login.defs")
+    lines.find { |line| line.start_with?("GID_MAX") }.split("\s")[1].strip
```
What happens if the line looks like this:

```
GID_MAX        50000
```

That is if there are multiple spaces between GID_MAX and the value? Will \[1\] resolve as " " (=a space)?

Also, what if there are multiple lines that start with GID_MAX? Normally that would not happen, but can't be ruled out. That would cause potentially hard to debug problems. Two options I can think of for handling that:

* Error out with a reasonable error message ("ERROR: multiple GID_MAX lines found in /etc/login.defs")
* Pick the first value (e.g. *line.starts_with?("GID_MAX")[0]*) and print an informal message ("WARNING: multiple GIT_MAX lines found in /etc/login.defs").

mattock replied:
Based on testing this works as expected. The "find" method returns the first line that matches. The split("\s") considers multiple whitespace characters as one, so multiple spaces between GID_MAX and the value do not cause issues.

gartemiev commented:
@mattock , can this be fixed regarding tests and merged?

mattock commented:
> @mattock , can this be fixed regarding tests and merged?

@kibahop needs to be pushed a bit.

mattock commented:
I think this is mergeable once the commits have been squashed.

gartemiev commented:
@mattock , could you please squash it? I don't have permissions for it: 

`
remote: Permission to Puppet-Finland/puppet-ipa.git denied to gartemiev.
fatal: unable to access 'https://github.com/Puppet-Finland/puppet-ipa.git/': The requested URL returned error: 403
`

To quash it you need to run: 
`git rebase -i HEAD~16`

and then add squash to all commits except initial one.

kibahop commented:
Add support for RHEL9

gartemiev commented:
@mattock , this is still an issue with merge - commits were squashed by @kibahop - can you please fix and merge it?

kibahop committed:
Bring documentation up to date

Signed-off-by: Kibahop <petri.lammi@puppeteers.net>

kibahop committed:
pdk: update from tags/2.5.0 to 2.7.4

Signed-off-by: Kibahop <petri.lammi@puppeteers.net>

mattock committed:
Autocorrect linting issues

kibahop committed:
Update docs and REFERENCE.md

Signed-off-by: Kibahop <petri.lammi@puppeteers.net>

kibahop committed:
More docs cleanups.

Signed-off-by: Kibahop <petri.lammi@puppeteers.net>

mattock committed:
Fix GitHub Actions

The old workflow was from voxpupuli and was not really PDK-compatible. It could
be made to work only by heavy modifications to the Gemfile, which made it
practically incompatible with "pdk update" and .sync.yml. So, use a
PDK-compatible workflow from puppetlabs-postgresql with trivial modifications.

Signed-off-by: Samuli Seppänen <samuli@openvpn.net>

gartemiev commented:
@mattock - thanks for fixing tests!!! Highly appreciate it!!!
